### DISCUSSION TITLE
Issue with masking

### DISCUSSION THREAD

**Discussion by scottyv80** [roflcoopter/viseron#483] [Q&A]
First, thank you for this.  I have been running it for a couple of weeks and have liked what I have seen.

I am having an issue with masking on object detection. I have used image-map.net to get the coordinates for the masking of two stationary objects that keep being incorrectly identified as dogs.  I entered a mask and they were still being constantly detected.  I increased the mask size and they are still being identified as a dog.  The objects are a landscaping spotlight and a lid to an underground propane tank.  Is there a minimum size to the mask? Are there any suggestions to get the masks to work correctly.  I am using Darknet and Mog2.

**Comment by roflcoopter** [comment]
Can you double check if the mask is set where you expect it to be?
You can check by going to the MJPEG stream and enabling drawing of masks, explained here.
Just realized the documentation does not mention `draw_object_mask`

It will look something like this:
``` 
http://localhost:8888/<camera name slug>/mjpeg-stream?draw_object_mask`
```

**Reply by scottyv80** [reply]
When looking at the MJPEG stream and the draw_object_mask, I am not seeing any of my masks that I have entered.  I have added four masks to the configuration file.  I took the code from the site and did and alternated x and y with the codes.  Where should I go from here.

Thanks!

**Reply by roflcoopter** [reply]
Could you post your config?

**Reply by scottyv80** [reply]
Certainly.  Thank you!


## Cameras
ffmpeg:
  camera:
    camera_1:
      name: Front Door
      host: {camera ip address}
      port: 554
      path: rtsp://{path to stream}/cam/realmonitor?channel=1&subtype=0
      stream_format: rtsp
      username: {username}
      password: {password}

## Object detector
darknet:
  object_detector:
    cameras:
      camera_1:
        fps: 1
        scan_on_motion_only: false
        labels:
          - label: person
            confidence: 0.75
            trigger_recorder: true
            require_motion: true
          - label: dog
            confidence: 0.50
            trigger_recorder: true
          - label: car
            confidence: 0.75
            trigger_recorder: true
            require_motion: true
          - label: truck
            confidence: 0.75
            trigger_recorder: true
            require_motion: true
        mask:
          - coordinates: #top of screen rectangle
              - x: 1
                y: 1
              - x: 2451
                y: 797
              - x: 2451
                y: 1
              - x: 1
                y: 797
          - coordinates: #Propane Tank rectangle
              - x: 1453
                y: 1034
              - x: 1453
                y: 1120
              - x: 1568
                y: 1034
              - x: 1568
                y: 1120
          - coordinates: #Spot Light Poly
              - x: 998
                y: 1116
              - x: 1070
                y: 1118
              - x: 1257
                y: 1254
              - x: 1231
                y: 1309
              - x: 1174
                y: 1307
              - x: 971
                y: 1167
          - coordinates:
              - x: 252
                y: 827
              - x: 337
                y: 890
              - x: 252
                y: 890
              - x: 337
                y: 827

## Motion detection
mog2:
  motion_detector:
    cameras:
      camera_1:
        trigger_recorder: true
        fps: 3
        mask:
          - coordinates: #top of screen rectangle
              - x: 1
                y: 1
              - x: 2451
                y: 797
              - x: 2451
                y: 1
              - x: 1
                y: 797
          - coordinates: #Propane Tank rectangle
              - x: 1453
                y: 1034
              - x: 1453
                y: 1120
              - x: 1568
                y: 1034
              - x: 1568
                y: 1120
          - coordinates: #Spot Light Poly
              - x: 998
                y: 1116
              - x: 1070
                y: 1118
              - x: 1257
                y: 1254
              - x: 1231
                y: 1309
              - x: 1174
                y: 1307
              - x: 971
                y: 1167
          - coordinates:
              - x: 252
                y: 827
              - x: 337
                y: 890
              - x: 252
                y: 890
              - x: 337
                y: 827
        area: 0.04
        threshold: 15
        history: 250

nvr:
  camera_1:

mqtt:
  broker: {Mqtt broker ip}
  port: {port}
  username: {username}
  password: {password}
  client_id: viseron
  home_assistant:

**Reply by roflcoopter** [reply]
Can you post it in a code block so the formatting is kept?

This article shows how: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by scottyv80** [reply]
Sorry, still learning github.  Let's try this.
```
## Cameras
ffmpeg:
  camera:
    camera_1:
      name: Front Door
      host: {camera IP}
      port: 554
      path: rtsp:{camera IP}/cam/realmonitor?channel=1&subtype=0
      stream_format: rtsp
      username: {username}
      password: {password}
      
## Object detector
darknet:
  object_detector:
    cameras:
      camera_1:
        fps: 1
        scan_on_motion_only: false  
        labels:
          - label: person
            confidence: 0.75
            trigger_recorder: true
            require_motion: true
          - label: dog
            confidence: 0.50
            trigger_recorder: true
#            require_motion: true
          - label: car
            confidence: 0.75
            trigger_recorder: true
            require_motion: true
          - label: truck
            confidence: 0.75
            trigger_recorder: true
#            require_motion: true
        mask:
          - coordinates: #top of screen rectangle
              - x: 1
                y: 1
              - x: 2451
                y: 797
              - x: 2451
                y: 1
              - x: 1
                y: 797
          - coordinates: #Propane Tank rectangle
              - x: 1453
                y: 1034
              - x: 1453
                y: 1120
              - x: 1568
                y: 1034
              - x: 1568
                y: 1120
          - coordinates: #Spot Light Poly
              - x: 998
                y: 1116
              - x: 1070
                y: 1118
              - x: 1257
                y: 1254
              - x: 1231
                y: 1309
              - x: 1174
                y: 1307
              - x: 971
                y: 1167
          - coordinates:
              - x: 252
                y: 827
              - x: 337
                y: 890
              - x: 252
                y: 890
              - x: 337
                y: 827

## Motion detection
mog2:
  motion_detector:
    cameras:
      camera_1:
        trigger_recorder: true
        fps: 3
        mask:
          - coordinates: #top of screen rectangle
              - x: 1
                y: 1
              - x: 2451
                y: 797
              - x: 2451
                y: 1
              - x: 1
                y: 797
          - coordinates: #Propane Tank rectangle
              - x: 1453
                y: 1034
              - x: 1453
                y: 1120
              - x: 1568
                y: 1034
              - x: 1568
                y: 1120
          - coordinates: #Spot Light Poly
              - x: 998
                y: 1116
              - x: 1070
                y: 1118
              - x: 1257
                y: 1254
              - x: 1231
                y: 1309
              - x: 1174
                y: 1307
              - x: 971
                y: 1167
          - coordinates:
              - x: 252
                y: 827
              - x: 337
                y: 890
              - x: 252
                y: 890
              - x: 337
                y: 827
        area: 0.04
        threshold: 15
        history: 250

nvr:
  camera_1:

mqtt:
  broker: {mqtt IP address}
  port: {mqtt port}
  username: {mqtt username}
  password: {mqtt password}
  client_id: {client_id}
  home_assistant:

**Reply by roflcoopter** [reply]
Thanks!

And also, what's your cameras resolution?

**Reply by scottyv80** [reply]
The main stream is 2592x1944.

**Reply by roflcoopter** [reply]
Sorry i hade a mistake in the URL to the MJPEG stream.
It should be:
```
http://localhost:8888/<camera name slug>/mjpeg-stream?draw_object_mask=1
```

Can you see the mask with that URL?

**Reply by scottyv80** [reply]
I see this on the browser: 500: Internal Server Error

The docker terminal shows this message when I went to the URL:
[2023-02-21 22:03:40] [ERROR   ] [tornado.application] - Uncaught exception GET /camera_1/mjpeg-stream?draw_object_mask=1 ({IP_address})
HTTPServerRequest(protocol='http', host='{IP_address}:8888', method='GET', uri='/camera_1/mjpeg-stream?draw_object_mask=1', version='HTTP/1.1', remote_ip='{IP_address}')
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tornado/web.py", line 1704, in _execute
    result = await result
  File "/src/viseron/components/webserver/stream_handler.py", line 149, in get
    ret, jpg = await self.process_frame(
  File "/src/viseron/components/webserver/stream_handler.py", line 83, in process_frame
    draw_object_mask(
  File "/src/viseron/helpers/__init__.py", line 271, in draw_object_mask
    draw_mask("Object mask", frame, mask_points, color=(255, 255, 255))
  File "/src/viseron/helpers/__init__.py", line 245, in draw_mask
    center_x = int(image_moment["m10"] / image_moment["m00"])
ZeroDivisionError: float division by zero
[2023-02-21 22:03:40] [ERROR   ] [tornado.access] - 500 GET /camera_1/mjpeg-stream?draw_object_mask=1 ({IP_address}) 295.76ms

**Comment by roflcoopter** [comment]
Hmm, and `rtsp:{camera IP}/cam/realmonitor?channel=1&subtype=0` points to the main stream of your camera?

It appears that the center X coordinate of your mask is outside of the frame, will need to do some more digging and maybe skip adding the text that is causing that exception

**Comment by roflcoopter** [comment]
I just pushed a fix to the `dev` tag if you wanna try it out.
Seems that your mask might not be closed properly, thats why it errors out.

You should be able to see that drawn mask now which may give you some clues
